fendermint/app/src/app.rs

```diff
@@ -0,0 +1,299 @@
+    {
+        let state = self.take_exec_state();
+        let (state, ret) = f(state).await?;
+        self.put_exec_state(state);
+        Ok(ret)
+    }
+}
+
+// NOTE: The `Application` interface doesn't allow failures at the moment. The protobuf
+// of `Response` actually has an `Exception` type, so in theory we could use that, and
+// Tendermint would break up the connection. However, before the response could reach it,
+// the `tower-abci` library would throw an exception when it tried to convert a
+// `Response::Exception` into a `ConensusResponse` for example.
+#[async_trait]
+impl<DB, I> Application for App<DB, I>
+where
+    DB: Blockstore + Clone + Send + Sync + 'static,
+    I: Interpreter<
+        State = FvmState<DB>,
+        Message = Vec<u8>,
+        BeginOutput = (),
+        DeliverOutput = BytesMessageApplyRet,
+        EndOutput = (),
+    >,
+{
+    /// Provide information about the ABCI application.
+    async fn info(&self, _request: request::Info) -> response::Info {
+        let state = self.committed_state();
+        let height =
+            tendermint::block::Height::try_from(state.block_height).expect("height too big");
+        let app_hash = tendermint::hash::AppHash::try_from(state.state_root.to_bytes())
+            .expect("hash can be wrapped");
+        response::Info {
+            data: "fendermint".to_string(),
+            version: VERSION.to_owned(),
+            app_version: 1,
+            last_block_height: height,
+            last_block_app_hash: app_hash,
+        }
+    }
+
+    /// Called once upon genesis.
+    async fn init_chain(&self, _request: request::InitChain) -> response::InitChain {
+        Default::default()
+    }
+
+    /// Query the application for data at the current or past height.
+    async fn query(&self, _request: request::Query) -> response::Query {
+        todo!("make a query interpreter")
+    }
+
+    /// Check the given transaction before putting it into the local mempool.
+    async fn check_tx(&self, _request: request::CheckTx) -> response::CheckTx {
+        todo!("make an interpreter for checks, on a projected state")
+    }
+
+    /// Signals the beginning of a new block, prior to any `DeliverTx` calls.
+    async fn begin_block(&self, request: request::BeginBlock) -> response::BeginBlock {
+        let state = self.committed_state();
+        let height = request.header.height.into();
+        let timestamp = Timestamp(
+            request
+                .header
+                .time
+                .unix_timestamp()
+                .try_into()
+                .expect("negative timestamp"),
+        );
+        let db = self.db.as_ref().to_owned();
+
+        let state = FvmState::new(
+            db,
+            height,
+            timestamp,
+            state.network_version,
+            state.state_root,
+            state.base_fee,
+            state.circ_supply,
+        )
+        .expect("error creating new state");
```

Comment:
Yes, that's right, the `Application` trait won't let me do otherwise. Today I thought I'd change it; it's only like this because the `tendermint-rs` project has a similar trait (just synchronous). It was created in https://github.com/consensus-shipyard/fendermint/issues/3 

However, the way the `tower-abci` I'm using to host the service works is that it would panic if I returned the `Response::Exception` kind that Tendermint allows. Even if I did, the result would be Tendermint dropping the connection [here](https://github.com/tendermint/tendermint/blob/64747b2b184184ecba4f4bffc54ffbcb47cfbcb0/abci/client/socket_client.go#L166), and then we'd have an application nobody talks to any more. 

So yeah, panicking here will kill the application, which will show up in the Tendermint logs as well because the connections will be lost.